aiidalab/app.py

```diff
@@ -82,14 +82,131 @@ class AiidaLabApp(traitlets.HasTraits):
+
+            self.commit = match.groupdict()['commit']
+            self.short_ref = match.groupdict()['short_ref']
+            assert self.commit or self.short_ref
+
+        @property
+        def _repo(self):
+            return Repo(self.app.path)
+
+        def _resolve_short_ref(self, short_ref):
+            """Attempt to resolve the short-ref to a full ref.
+
+            For example, 'branch' would be resolved to 'refs/heads/branch'
+            if 'branch' is a local branch or 'refs/tags/branch' if it was
+            a tag.
+
+            This function returns None if the short-ref cannot be resolved
+            to a full reference.
+            """
+            # Check if short-ref is a head:
+            if f'refs/heads/{short_ref}'.encode() in self._repo.refs.allkeys():
+                return f'refs/heads/{short_ref}'.encode()
+
+            # Check if short-ref is a tag:
+            if f'refs/tags/{short_ref}'.encode() in self._repo.refs.allkeys():
+                return f'refs/tags/{short_ref}'.encode()
+
+            # Check if short-ref is among the remote refs:
+            for ref in self._repo.refs.allkeys():
+                if re.match(r'refs\/remotes\/(.*)?\/' + short_ref, ref.decode()):
+                    return ref
+
+            return None
+
+        def find_versions(self):
+            """Find versions available for this release line.
+
+            When encountering an ambiguous release line name, i.e.,
+            a shared branch and tag name, we give preference to the
+            branch, because that is what git does in this situation.
+            """
+            assert self.short_ref or self.commit
+
+            if self.commit:  # The release line is a commit.
+                assert self.commit.encode() in self._repo.object_store
+                yield self.commit.encode()
+            else:
+                ref = self._resolve_short_ref(self.short_ref)
+                if ref is None:
+                    raise ValueError(f"Unable to resolve '{self.short_ref}'. "
+                                     "Are you sure this is a valid git branch or tag?")
+
+                # The release line is a head (branch).
+                if ref.startswith(b'refs/heads/'):
+                    ref_commit = self._repo.get_peeled(ref)
+                    all_tags = {
+                        ref: c for ref, c in self._repo.refs.as_dict().items() if ref.startswith('refs/tags'.encode())
+                    }
+                    tags_lookup = {c: ref for ref, c in all_tags.items()}
+                    tagged_commits_on_head = [
+                        c.commit.id for c in self._repo.get_walker(self._repo.refs[ref]) if c.commit.id in tags_lookup
+                    ]
+                    yield tags_lookup.get(ref_commit, ref)
+                    for commit in tagged_commits_on_head:
+                        if commit != ref_commit:
+                            yield tags_lookup[commit]
```

Comment:
Is it supposed to yield the first one here?

Reply:
You mean on line 174? Yes.

Reply:
So the logic is that the first thing returned is the `ref_commit` and all subsequent returned objects are `tagged_commits_on_head`?

Reply:
Yes, exactly.